h2o-algos/src/main/java/hex/rulefit/Rule.java

```diff
@@ -85,43 +86,57 @@ public boolean equals(Object obj) {
     
     public static Set<Rule> extractRulesFromTree(SharedTreeSubgraph tree, int modelId, String classString) {
         Set<Rule> rules = new HashSet<>();
-        List<Condition> conditions = new ArrayList<>();
-        traverseNodes(tree.rootNode, conditions, rules, null, modelId, classString);
-        return rules;
-    }
-    
-    private static void traverseNodes(SharedTreeNode node, List<Condition> conditions, Set<Rule> rules, Condition conditionToAdd, int modelId, String classString) {
-        if (conditionToAdd != null) {
-            conditions.add(conditionToAdd);
-        }
-        
-        if (node.isLeaf()) {
-            // create Rule
-            String varName = "M" + modelId + "T" + node.getSubgraphNumber() + "N" + node.getNodeNumber();
+        // filter leaves
+        List<SharedTreeNode> leaves = tree.nodesArray.stream().filter(sharedTreeNode -> sharedTreeNode.isLeaf()).collect(Collectors.toList());
+        // traverse paths
+        for (SharedTreeNode leaf : leaves) {
+            List<Condition> conditions = new ArrayList<>();
```

Comment:
it would be a little clearer if there was a version fo the `traversePath` method that doesn't need the empty array on the input

```
void traversePath(SharedTreeNode node, Set<Rule> rules, String varName) {
   traversePath(node, new ArrayList<>(), rules, String varName)
}
```

Reply:
thanks, abef4ad